log4j-core/src/main/java/org/apache/logging/log4j/core/util/UuidUtil.java

```diff
@@ -140,4 +142,47 @@ public static UUID getTimeBasedUuid() {
         final long most = timeLow | timeMid | TYPE1 | timeHi;
         return new UUID(most, LEAST);
     }
+
+    /**
+     * Generates a Type 4 UUID based on the deterministic LogEvent hash.
+     * Meant for generating consistent, correlatable UUID values across multiple Appenders for the same LogEvent.
+     *
+     * @param logEvent
+     * @return universally unique identifiers (UUID)
+     */
+    public static UUID getLogEventBasedUuid(LogEvent logEvent) {
+        // TODO: better hashing algorithm - include other LogEvent fields?
+        long epochSecond = logEvent.getInstant().getEpochSecond();
+        // Enable 'log4j.configuration.usePreciseClock' system property otherwise will be truncated to millis
+        long nanoOfSecond = logEvent.getInstant().getNanoOfSecond();
+        // Thread IDs typically increment from 0 producing a narrow range
+        long threadId = logEvent.getThreadId();
+
+        // Increase entropy
+        long most = mix(epochSecond, nanoOfSecond, threadId);
+        long least = mix(threadId, ~nanoOfSecond, epochSecond);
+        // Set UUID v4 bits
+        most &= 0xFFFFFFFFFFFF0FFFL;
+        most |= 0x0000000000004000L;
```

Comment:
This looks more like a [UUID Type 8](https://www.rfc-editor.org/rfc/rfc9562.html#name-uuid-version-8) (vendor-specific or experimental), not a UUID Type 4 (purely random).